### PR TITLE
v08 - fix ContinueAs data field assignment

### PR DESCRIPTION
**Pull request by spolti**
Signed-off-by: spolti <filippespolti@gmail.com>

**Many thanks for submitting your Pull Request :heart:!**

**What this PR does / why we need it**:

**Special notes for reviewers**:

**Additional information (if needed):**

### CODE REVIEW

File: model/workflow.go [hunk 1]
```diff
@@ -206,6 +206,14 @@ func (w *Workflow) setDefaults() {
 	}
 }
 
+// WorkflowRef holds a reference for a workflow definition
+type WorkflowRef struct {
```
Comment (lsytj0413) [inline model/workflow.go:210]: This struct is used by Action & Workflow struct, so should we place it in workflow_ref.go file?

File: model/workflow.go [hunk 2]
```diff
@@ -486,12 +494,45 @@ func (e *End) UnmarshalJSON(data []byte) error {
...
 	// TODO: add object or string data type
-	// If string type, an expression which selects parts of the states data output to become the workflow data input of continued execution. If object type, a custom object to become the workflow data input of the continued execution
+	// If string type, an expression which selects parts of the states data output to become the workflow data input of
+	// continued execution. If object type, a custom object to become the workflow data input of the continued execution
 	Data interface{} `json:"data,omitempty"`
 	// WorkflowExecTimeout Workflow execution timeout to be used by the workflow continuing execution. Overwrites any specific settings set by that workflow
 	WorkflowExecTimeout WorkflowExecTimeout `json:"workflowExecTimeout,omitempty"`
 }
 
+func (c *ContinueAs) UnmarshalJSON(data []byte) error {
```
Comment (lsytj0413) [inline model/workflow.go:504]: Should we change this style unmarshal to https://github.com/serverlessworkflow/sdk-go/blob/main/model/state_exec_timeout.go#L35 style？

File: model/workflow.go (hunk 1 above)
Comment (spolti) [reply model/workflow.go:210]: it will have just one thing in there, maybe later as it grows?

Comment (spolti) [reply model/workflow.go:210]: maybe after we get your PR merged.

File: model/workflow.go (hunk 2 above)
Comment (spolti) [reply model/workflow.go:504]: tried something similar, not sure why it doesn't set the data and the timeout values.

File: model/workflow.go [hunk 3]
```diff
@@ -501,36 +501,31 @@ type ContinueAs struct {
 	WorkflowExecTimeout WorkflowExecTimeout `json:"workflowExecTimeout,omitempty"`
 }
 
+//type testFOrUnmarshall ContinueAs
+
 func (c *ContinueAs) UnmarshalJSON(data []byte) error {
-	rewf := make(map[string]interface{})
-	if err := json.Unmarshal(data, &rewf); err != nil {
+	continueAs := make(map[string]json.RawMessage)
```
Comment (spolti) [inline model/workflow.go:507]: @lsytj0413 or something like this.

Comment (lsytj0413) [reply model/workflow.go:507]: I don't know what the different of previous implement，it always unmarshal multi times right？
There is some explain how StateExecTimeout style works https://github.com/serverlessworkflow/sdk-go/pull/76#issuecomment-1253100748

**Comment by spolti** [comment]
relates to https://github.com/serverlessworkflow/sdk-go/issues/70

File: model/workflow.go (hunk 3 above)
Comment (spolti) [reply model/workflow.go]: did a small update, please take a look.

**Comment by spolti** [comment]
interesting, I requested @lsytj0413  to re-review and it remove you @ricardozanini lol

**Review by ricardozanini — APPROVED** [review]

File: model/workflow.go [hunk 4]
```diff
@@ -15,8 +15,12 @@
 package model
 
 import (
+	"bytes"
 	"encoding/json"
 	"fmt"
+	"github.com/go-playground/validator/v10"
+	val "github.com/serverlessworkflow/sdk-go/v2/validator"
```
Comment (lsytj0413) [inline model/workflow.go:22]: should we organize the imports by it's category？

File: model/workflow.go [hunk 5]
```diff
@@ -47,6 +51,20 @@ var actionsModelMapping = map[string]func(state map[string]interface{}) State{
...
+	val.GetValidator().RegisterStructValidation(continueAsStructLevelValidation, ContinueAs{})
+}
+
+func continueAsStructLevelValidation(structLevel validator.StructLevel) {
+	continueAs := structLevel.Current().Interface().(ContinueAs)
+	if len(continueAs.WorkflowExecTimeout.Duration) > 0 {
+		if err := val.ValidateISO8601TimeDuration(continueAs.WorkflowExecTimeout.Duration); err != nil {
+			structLevel.ReportError(reflect.ValueOf(continueAs.WorkflowExecTimeout.Duration),
+				"workflowExecTimeout", "dration", "reqiso8601duration", "")
+		}
```
Comment (lsytj0413) [inline model/workflow.go:64]: how about use `iso8601duration` tag？

File: model/workflow.go (hunk 4 above)
Comment (spolti) [reply model/workflow.go:22]: it was edited by intellij, we can later define a code style and adjust it.

File: model/workflow.go [hunk 6]
```diff
@@ -482,16 +500,73 @@ func (e *End) UnmarshalJSON(data []byte) error {
...
 	// TODO: add object or string data type
-	// If string type, an expression which selects parts of the states data output to become the workflow data input of continued execution. If object type, a custom object to become the workflow data input of the continued execution
+	// If string type, an expression which selects parts of the states data output to become the workflow data input of
+	// continued execution. If object type, a custom object to become the workflow data input of the continued execution
 	Data interface{} `json:"data,omitempty"`
 	// WorkflowExecTimeout Workflow execution timeout to be used by the workflow continuing execution. Overwrites any specific settings set by that workflow
 	WorkflowExecTimeout WorkflowExecTimeout `json:"workflowExecTimeout,omitempty"`
 }
 
+type continueAsForUnmarshall ContinueAs
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go:514]: yes :)

File: model/workflow.go [hunk 7]
```diff
@@ -482,16 +500,73 @@ func (e *End) UnmarshalJSON(data []byte) error {
...
+
+	var err error
+	switch data[0] {
+	case '"':
+		c.WorkflowID, err = unmarshalString(data)
+		if err != nil {
+			return err
+		}
+		c.Invoke, c.OnParentComplete = "sync", "terminate"
+		return nil
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go:531]: if I don't to this, the WorkflowRef fields will not be set..
e.g:
```
                Error Trace:    parser_test.go:357
                Error:          Received unexpected error:
                                Key: 'Workflow.States[1].DataConditions[0].End.ContinueAs.WorkflowRef.WorkflowID' Error:Field validation for 'WorkflowID' failed on the 'required' tag

}
```

File: model/workflow.go [hunk 8]
```diff
@@ -482,16 +500,73 @@ func (e *End) UnmarshalJSON(data []byte) error {
...
+			}
+			return nil
+		}
+
+		if err := unmarshalKey("data", continueAs, &c.Data); err != nil {
+			return err
+		}
+		if err := unmarshalKey("workflowExecTimeout", continueAs, &c.WorkflowExecTimeout); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go:547]: not really, the unmarshal below will not set the fields as well.

File: model/workflow.go (hunk 5 above)
Comment (spolti) [reply model/workflow.go:64]: can be, will replace it where it occurs.

File: model/workflow.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lsytj0413) [reply model/workflow.go:522]: Ok, seems the behavior is different with embedded struct field.

File: model/workflow.go [hunk 9]
```diff
@@ -58,6 +61,20 @@ var actionsModelMapping = map[string]func(state map[string]interface{}) State{
...
+func init() {
+	val.GetValidator().RegisterStructValidation(continueAsStructLevelValidation, ContinueAs{})
+}
+
+func continueAsStructLevelValidation(structLevel validator.StructLevel) {
+	continueAs := structLevel.Current().Interface().(ContinueAs)
+	if len(continueAs.WorkflowExecTimeout.Duration) > 0 {
+		if err := val.ValidateISO8601TimeDuration(continueAs.WorkflowExecTimeout.Duration); err != nil {
+			structLevel.ReportError(reflect.ValueOf(continueAs.WorkflowExecTimeout.Duration),
+				"workflowExecTimeout", "dration", "iso8601duration", "")
```
Comment (lsytj0413) [inline model/workflow.go:73]: dration -> duration？

Comment (spolti) [reply model/workflow.go:73]: probably, copy and paste issue :D

**Pull request merged**